Arrayoutofbounds exception

Note that due to the vagaries of 2's complement notation, Math.abs(Integer.MIN_VALUE) == Integer.MIN_VALUE

com.tc.exception.TCRuntimeException: Uncaught exception in stage
        at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:211)
Caused by: java.lang.ArrayIndexOutOfBoundsException: -58
        at com.tc.async.impl.StageQueueImpl.findShortestQueueIndex(StageQueueImpl.java:208)
        at com.tc.async.impl.StageQueueImpl.getSourceQueueFor(StageQueueImpl.java:224)
        at com.tc.async.impl.StageQueueImpl.addMultiThreaded(StageQueueImpl.java:152)
        at com.tc.objectserver.entity.RequestProcessor.scheduleRequest(RequestProcessor.java:101)
        at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.start(ManagedEntityImpl.java:1040)
        at com.tc.objectserver.entity.ManagedEntityImpl$SchedulingRunnable.access$100(ManagedEntityImpl.java:1001)
        at com.tc.objectserver.entity.ManagedEntityImpl.scheduleInOrder(ManagedEntityImpl.java:291)
        at com.tc.objectserver.entity.ManagedEntityImpl.processInvokeRequest(ManagedEntityImpl.java:319)
        at com.tc.objectserver.entity.ManagedEntityImpl.addRequestMessage(ManagedEntityImpl.java:199)
        at com.tc.objectserver.handler.ProcessTransactionHandler.addMessage(ProcessTransactionHandler.java:294)
        at com.tc.objectserver.handler.ProcessTransactionHandler.access$400(ProcessTransactionHandler.java:75)
        at com.tc.objectserver.handler.ProcessTransactionHandler$2.handleEvent(ProcessTransactionHandler.java:141)
        at com.tc.objectserver.handler.ProcessTransactionHandler$2.handleEvent(ProcessTransactionHandler.java:122)
        at com.tc.async.impl.StageQueueImpl$HandledContext.runWithHandler(StageQueueImpl.java:513)
        at com.tc.async.impl.StageImpl$WorkerThread.run(StageImpl.java:197)

--- a/common/src/main/java/com/tc/async/impl/StageImpl.java
+++ b/common/src/main/java/com/tc/async/impl/StageImpl.java
@@ -57,12 +57,7 @@ public class StageImpl<EC> implements Stage<EC> {
    * @param loggerProvider : logger
    * @param name : The stage name
    * @param handler : Event handler for this stage
-   * @param threadCount : Number of threads working on this stage
-   * @param threadsToQueueRatio : The ratio determines the number of queues internally used and the number of threads
-   *        per each queue. Ideally you would want this to be same as threadCount, in which case there is only 1 queue
-   *        used internally and all thread are working on the same queue (which doesn't guarantee order in processing)
-   *        or set it to 1 where each thread gets its own queue but the (multithreaded) event contexts are distributed
-   *        based on the key they return.
+   * @param queueCount : Number of threads and queues working on this stage with 1 thread bound to 1 queue
    * @param group : The thread group to be used
    * @param queueFactory : Factory used to create the queues
    * @param queueSize : Max queue Size allowed

--- a/common/src/main/java/com/tc/async/impl/StageQueueImpl.java
+++ b/common/src/main/java/com/tc/async/impl/StageQueueImpl.java
@@ -46,15 +46,11 @@ public class StageQueueImpl<EC> implements Sink<EC> {
   private final TCLogger          logger;
   private final SourceQueueImpl<ContextWrapper<EC>>[] sourceQueues;
   private volatile boolean closed = false;
+  private volatile int fcheck = 0;  // used to start the shortest queue search
   /**
    * The Constructor.
    * 
-   * @param threadCount : Number of threads working on this stage
-   * @param threadsToQueueRatio : The ratio determines the number of queues internally used and the number of threads
-   *        per each queue. Ideally you would want this to be same as threadCount, in which case there is only 1 queue
-   *        used internally and all thread are working on the same queue (which doesn't guarantee order in processing)
-   *        or set it to 1 where each thread gets its own queue but the (multithreaded) event contexts are distributed
-   *        based on the key they return.
+   * @param queueCount : Number of queues working on this stage
    * @param queueFactory : Factory used to create the queues
    * @param loggerProvider : logger
    * @param stageName : The stage name
@@ -70,17 +66,6 @@ public class StageQueueImpl<EC> implements Sink<EC> {
     createWorkerQueues(queueCount, queueFactory, queueSize, stageName);
   }
   
-  @SuppressWarnings("unchecked")  // for tests
-  StageQueueImpl(int queueCount, QueueFactory<ContextWrapper<EC>> queueFactory,
-                        TCLoggerProvider loggerProvider, String stageName, int queueSize, int fc) {
-    Assert.eval(queueCount > 0);
-    this.logger = loggerProvider.getLogger(Sink.class.getName() + ": " + stageName);
-    this.stageName = stageName;
-    this.sourceQueues = new SourceQueueImpl[queueCount];
-    this.fcheck = fc;
-    createWorkerQueues(queueCount, queueFactory, queueSize, stageName);
-  }
-  
   private void createWorkerQueues(int queueCount, QueueFactory<ContextWrapper<EC>> queueFactory, int queueSize, String stage) {
     StageQueueStatsCollector statsCollector = new NullStageQueueStatsCollector(stage);
     BlockingQueue<ContextWrapper<EC>> q = null;
@@ -92,7 +77,7 @@ public class StageQueueImpl<EC> implements Sink<EC> {
 
     for (int i = 0; i < queueCount; i++) {
       q = queueFactory.createInstance(queueSize);
-      this.sourceQueues[i] = new SourceQueueImpl<ContextWrapper<EC>>(q, String.valueOf(queueCount), statsCollector);
+      this.sourceQueues[i] = new SourceQueueImpl<ContextWrapper<EC>>(q, i, statsCollector);
     }
   }
 
@@ -193,29 +178,27 @@ public class StageQueueImpl<EC> implements Sink<EC> {
     }
   }
   
-  private volatile int fcheck = 0;
-//  TODO:  Way too busy. need a better way
+//  TODO:  Way too busy. REALLY need a better way
   private int findShortestQueueIndex() {
-      int stop = Math.abs(fcheck++) % this.sourceQueues.length;
-      int pointer = stop+1;
-      int min = Integer.MAX_VALUE;
-      int can = -1;
+    final int pointer = fcheck;
+    int min = Integer.MAX_VALUE;
+    int can = -1;
 // special case where context can go to any queue, pick the shortest
-      while (pointer != stop) {
-        if (++pointer >= this.sourceQueues.length) {
-          pointer = 0;
-        }
-        SourceQueueImpl<ContextWrapper<EC>> impl = this.sourceQueues[pointer];
-        if (impl.isEmpty()) {
-          return pointer;
-        } else {
-          if (Math.min(min, impl.size()) != min) {
-            can = pointer;
-            min = impl.size();
-          }
+    for (int x=0;x<this.sourceQueues.length;x++) {
+      int index = (pointer + x) % this.sourceQueues.length;
+      SourceQueueImpl<ContextWrapper<EC>> impl = this.sourceQueues[index];
+      if (impl.isEmpty()) {
+        return index;
+      } else {
+        int checkMin = impl.size(); // concurrent access so just use current value.  this method is best efforts
+        if (Math.min(min, checkMin) != min) {
+          can = index;
+          min = checkMin;
         }
       }
-      return can;
+    }
+    Assert.assertTrue(can >= 0 && can < this.sourceQueues.length);
+    return can;
   }
 
   private int getSourceQueueFor(MultiThreadedEventContext context) {
@@ -321,15 +304,15 @@ public class StageQueueImpl<EC> implements Sink<EC> {
     this.sourceQueues[0].getStatsCollector().reset();
   }
 
-  private static final class SourceQueueImpl<W> implements Source<W> {
+  private final class SourceQueueImpl<W> implements Source<W> {
 
     private final BlockingQueue<W> queue;
-    private final String                      sourceName;
+    private final int                      sourceIndex;
     private volatile StageQueueStatsCollector statsCollector;
 
-    public SourceQueueImpl(BlockingQueue<W> queue, String sourceName, StageQueueStatsCollector statsCollector) {
+    public SourceQueueImpl(BlockingQueue<W> queue, int sourceIndex, StageQueueStatsCollector statsCollector) {
       this.queue = queue;
-      this.sourceName = sourceName;
+      this.sourceIndex = sourceIndex;
       this.statsCollector = statsCollector;
     }
 
@@ -364,6 +347,12 @@ public class StageQueueImpl<EC> implements Sink<EC> {
       W rv = this.queue.poll(timeout, TimeUnit.MILLISECONDS);
       if (rv != null) {
         this.statsCollector.contextRemoved();
+        if (queue.isEmpty()) {
+          // set the empty index for shortest queue in hopes of catching it on the first try
+          fcheck = this.sourceIndex;
+        }
+      } else {
+        fcheck = this.sourceIndex;
       }
       return rv;
     }
@@ -379,7 +368,7 @@ public class StageQueueImpl<EC> implements Sink<EC> {
 
     @Override
     public String getSourceName() {
-      return this.sourceName;
+      return Integer.toString(this.sourceIndex);
     }
   }
 

--- a/common/src/test/java/com/tc/async/impl/StageQueueImplTest.java
+++ b/common/src/test/java/com/tc/async/impl/StageQueueImplTest.java
@@ -34,10 +34,9 @@ import org.junit.Before;
 import org.junit.BeforeClass;
 import org.junit.Test;
 import static org.junit.Assert.*;
+import org.junit.Ignore;
 import org.mockito.Matchers;
-import org.mockito.Mockito;
 import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
 import org.mockito.invocation.InvocationOnMock;
 import org.mockito.stubbing.Answer;
@@ -136,7 +135,7 @@ public class StageQueueImplTest {
   }
   
   @Test
-  public void testShortestRollover() {
+  public void testShortestRollover() throws Exception {
     TCLoggerProvider logger = new DefaultLoggerProvider();
     final List<BlockingQueue<Object>> cxts = new ArrayList<BlockingQueue<Object>>();
     
@@ -151,13 +150,38 @@ public class StageQueueImplTest {
       }
     
     });
-    StageQueueImpl impl = new StageQueueImpl(1, context, logger, "mock", 16, Integer.MAX_VALUE);
+    StageQueueImpl impl = new StageQueueImpl(6, context, logger, "mock", 16);
     MultiThreadedEventContext cxt = mock(MultiThreadedEventContext.class);
     when(cxt.getSchedulingKey()).thenReturn(null);
- // hit int max
+ // fcheck starts at zero and should stay at zero because first queue is always empty
+    for (int x=0;x<6;x++) {
+      Assert.assertTrue(impl.getSource(0).isEmpty());
+      impl.addMultiThreaded(cxt);
+      Assert.assertNotNull(impl.getSource(0).poll(0));
+    }
+//  now try and fill one each on the the queues
+    for (int x=0;x<6;x++) {
+      Assert.assertTrue(impl.getSource(x).isEmpty());
+      impl.addMultiThreaded(cxt);
+      Assert.assertFalse(cxts.get(x).isEmpty());
+    }
+//  now clear the last three and re-fill them
+    for (int x=3;x<6;x++) {
+      Assert.assertFalse(impl.getSource(x).isEmpty());
+      Assert.assertNotNull(impl.getSource(x).poll(0));
+      Assert.assertTrue(cxts.get(x).isEmpty());
+      impl.addMultiThreaded(cxt);
+      Assert.assertFalse(cxts.get(x).isEmpty());
+    }
+//  now clear all again
+    for (int x=0;x<6;x++) {
+      Assert.assertFalse(impl.getSource(x).isEmpty());
+      Assert.assertNotNull(impl.getSource(x).poll(0));
+      Assert.assertTrue(cxts.get(x).isEmpty());
+    }
+// now add one more and make sure it is at the last queue since that was the 
+// last to be cleared
     impl.addMultiThreaded(cxt);
- // rollover
-    impl.addMultiThreaded(cxt);
-    Assert.assertEquals(cxts.get(0).size(), 2);
+    Assert.assertFalse(cxts.get(5).isEmpty());
   }
 }

